src/natcap/invest/urban_nature_access.py

```diff
@@ -617,63 +774,92 @@ def _greenspace_supply_demand_op(greenspace_budget, population):
-    valid_pixels = (convolved_population > 0)
-
-    # R_j is a ratio only calculated for the greenspace pixels.
-    greenspace_pixels = ~numpy.isclose(greenspace_area, 0.0)
-    valid_pixels &= greenspace_pixels
-    if population_nodata is not None:
-        valid_pixels &= ~numpy.isclose(
-            convolved_population, population_nodata)
-
-    if greenspace_nodata is not None:
-        valid_pixels &= ~numpy.isclose(greenspace_area, greenspace_nodata)
-
-    # If the population in the search radius is numerically 0, the model
-    # specifies that the ratio should be set to the greenspace area.
-    population_close_to_zero = numpy.isclose(convolved_population, 0.0)
-    out_array[population_close_to_zero] = (
-        greenspace_pixels[population_close_to_zero])
-    out_array[~greenspace_pixels] = 0.0  # set non-greenspace non-nodata to 0
-    out_array[valid_pixels] = (
-        greenspace_area[valid_pixels] / convolved_population[valid_pixels])
-
-    return out_array
+    greenspace_nodata = pygeoprocessing.get_raster_info(
+        greenspace_area_raster_path)['nodata'][0]
+    population_nodata = pygeoprocessing.get_raster_info(
+        convolved_population_raster_path)['nodata'][0]
+
+    def _greenspace_population_ratio(greenspace_area, convolved_population):
+        """Calculate the greenspace-population ratio R_j.
+
+        Args:
+            greenspace_area (numpy.array): A numpy array representing the area
+                of greenspace in the pixel.  Pixel values will be ``0`` if
+                there is no greenspace.  Pixel values may also match
+                ``greenspace_nodata``.
+            convolved_population (numpy.array): A numpy array where each pixel
+                represents the total number of people within a search radius of
+                each pixel, perhaps weighted by a search kernel.
+
+        Returns:
+            A numpy array with the ratio ``R_j`` representing the
+            greenspace-population ratio with the following constraints:
+
+                * ``convolved_population`` pixels that are numerically close to
+                  ``0`` are snapped to ``0`` to avoid unrealistically small
+                  denominators in the final ratio.
+                * Any non-greenspace pixels will have a value of ``0.0`` in the
+                  output matrix.
+        """
+        # ASSUMPTION: population nodata value is not close to 0.
+        #  Shouldn't be if we're coming from convolution.
+        out_array = numpy.full(
+            greenspace_area.shape, FLOAT32_NODATA, dtype=numpy.float32)
+
+        # Small negative values should already have been filtered out in
+        # another function after the convolution.
+        # This avoids divide-by-zero errors when taking the ratio.
+        valid_pixels = (convolved_population > 0)
+
+        # R_j is a ratio only calculated for the greenspace pixels.
+        greenspace_pixels = ~numpy.isclose(greenspace_area, 0.0)
```

Comment:
I think this and the other `0.0`s can just be `0` since python 3

Reply:
Yep, they sure don't: f8087c3ae